src/emitter.js

```diff
@@ -0,0 +1,279 @@
+		/**
+		 * Registers a callback function to be executed when an event is fired in a specific (emitter) object.
+		 *
+		 * @param {Emitter} emitter The object that fires the event.
+		 * @param {String} event The name of the event.
+		 * @param {Function} callback The function to be called on event.
+		 * @param {Object} [ctx] The object that represents `this` in the callback. Defaults to `emitter`.
+		 * @param {Number} [priority] The priority of this callback in relation to other callbacks to that same event.
+		 * Lower values are called first. Defaults to `10`.
+		 */
+		listenTo: function( emitter, event, callback, ctx, priority ) {
+			var emitters, emitterId, emitterInfo, eventCallbacks;
+
+			// _listeningTo contains a list of emitters that this object is listening to.
+			// This list has the following format:
+			//
+			// _listeningTo: {
+			//     emitterId: {
+			//         emitter: emitter,
+			//         callbacks: {
+			//             event1: [ callback1, callback2, ... ]
+			//             ....
+			//         }
+			//     },
+			//     ...
+			// }
+
+			if ( !( emitters = this._listeningTo ) ) {
+				emitters = this._listeningTo = {};
+			}
+
+			if ( !( emitterId = emitter._emitterId ) ) {
+				emitterId = emitter._emitterId = utils.uid();
+			}
+
+			if ( !( emitterInfo = emitters[ emitterId ] ) ) {
+				emitterInfo = emitters[ emitterId ] = {
+					emitter: emitter,
+					callbacks: {}
+				};
+			}
+
+			if ( !( eventCallbacks = emitterInfo.callbacks[ event ] ) ) {
+				eventCallbacks = emitterInfo.callbacks[ event ] = [];
+			}
+
+			eventCallbacks.push( callback );
+
+			// Finally register the callback to the event.
+			emitter.on( event, callback, ctx, priority );
+		},
+
+		/**
+		 * Stops listening for events. It can be usued at different levels:
+		 *
+		 *  * To stop listening to a specific callback.
+		 *  * To stop listening to a specific event.
+		 *  * To stop listening to all events fired by a specific object.
+		 *  * To stop listening to all events fired by all object.
+		 *
+		 * @param {Emitter} [emitter] The object to stop listening to. If omitted, stops it for all objects.
+		 * @param {String} [event] (Requires `emitter`) The name of the event to stop listening to. If omitted, stops it
+		 * for all events from `emitter`.
+		 * @param {Function} callback (Requires `event`) The function be removed from the call list for the give
+		 * `event`.
+		 */
+		stopListening: function( emitter, event, callback ) {
+			var emitters = this._listeningTo;
+			var emitterId = emitter && emitter._emitterId;
+			var emitterInfo = emitters && emitterId && emitters[ emitterId ];
+			var eventCallbacks = emitterInfo && event && emitterInfo.callbacks[ event ];
+
+			// Stop if nothing has been listened.
+			if ( !emitters || ( emitter && !emitterInfo ) || ( event && !eventCallbacks ) ) {
+				return;
+			}
+
+			if ( callback ) {
+				// All params provided. off() that single callback.
+				emitter.off( event, callback );
```

Comment:
> Therefore I think that they fit for the very same scenarios.

And what those scenarios are? I haven't worked with Backbone, but AFAIK 90% of cases are `obj.stopListening()`. If we are coding stuff which is not needed we are making a big mistake.

Besides that, I still don't like passing all the arguments to `off()` and even more to `stopListening()`, because not only this is long, but also problematic to do and maintain. If we make `on()` returning a listener instance (which is basically this [object](https://github.com/ckeditor/ckeditor5-core/pull/12/files#diff-d28ecdbfa507067c55c71c8d7c12167aR35) on steroids) we can IMO greatly improve the API.

> Anyway, those methods are already in place, paired with tests.

TBH, that should not be an argument. This way we will accept everything if it's already coded and working.

> They're not that complex... it's more about a series of if's checking parameters.

And these series are IMO unnecessarily complex. Plus, there's that big private object `this._listeningTo` which wouldn't be necessary if we resigned from the listeningTo's params and made `on()` return listener instance. We can make this cleaner and faster with little effort.